cli-output/src/cli_output.rs

```diff
@@ -66,28 +65,90 @@ pub enum OutputFormat {
     DisplayVerbose,
 }
 
-impl OutputFormat {
-    pub fn formatted_string<T>(&self, item: &T) -> String
-    where
-        T: Serialize + fmt::Display + QuietDisplay + VerboseDisplay,
-    {
-        match self {
-            OutputFormat::Display => format!("{item}"),
+/// CLI supports outputting processing results in the JSON format.  Objects that can be output in
+/// the JSON mode should implement this trait.
+///
+/// This trait is separate from, and is used by, the [`FormatForMode`] trait, for the convenience of
+/// implementation.  In most cases [`FormatForJson`] is implemented based on the [`Serialize`] trait
+/// that is automatically derived,  and an implementation for [`FormatForJson`] can then be produced
+/// by a macro.  On the other hand [`FormatForMode`] covers output for a human consumption.  And the
+/// later is often overwritten in a more nuanced way, so it can not be generated automatically.
+///
+/// This trait is also very similar to [`serde::Serialize`], but it is object safe, making it a bit
+/// more ergonomic in certain cases.  As we are producing whole `String`s, inlining of
+/// `format_for_json`, that monopolization can provide, most likely, have no real impact on the code
+/// performance.
+pub trait FormatForJson {
+    /// JSON output, but nicely formatted.
+    fn format_for_json(&self) -> serde_json::Result<String>;
+
+    /// JSON output, for machine parsing, as compact as possible.
+    fn format_for_json_compact(&self) -> serde_json::Result<String>;
+}
+
+/// Implements [`FormatForJson`] using [`serde_json::Serialize`].
+#[macro_export]
+macro_rules! impl_format_for_json_via_serialize {
+    ($type:ty) => {
+        impl $crate::FormatForJson for $type {
+            fn format_for_json(&self) -> $crate::reexport::serde_json::Result<String> {
+                $crate::reexport::serde_json::to_string_pretty(self)
+            }
+
+            fn format_for_json_compact(&self) -> $crate::reexport::serde_json::Result<String> {
+                $crate::reexport::serde_json::to_string(self)
+            }
+        }
+    };
+}
+
+/// A number of items we output to the CLI need to be formatted differently, depending on the
```

Comment:
What about 

```rust
/// Implement this trait for items that require varying formats in the CLI based on the [`OutputFormat`] mode.
```

Reply:
I think it is better to describe what something is, rather than specifying what you need to do with it.
In this particular case, the differences are rather minor.
But I fail to see why do you think your version is better.

As I wrote the description, it is obviously harder for me to understand what it means for someone to read it for the first time.

The only difference I see is that your version uses less words %)
I've rephrased mine to be shorter, but still follow the "what is it?" rather than "use it like this" description:

```rust
/// Describes how something can be converted into a string for a format specified by an
/// [`OutputFormat`] value.
pub trait FormatForMode: Display + FormatForJson {
```

Reply:
"Something" here is a type value for a type that implements the trait.
I do not think that the word "item" is a better choice here, to be honest.

"Something" here is a placeholder for a name of a type that you have at hand.
For example, say `CliAccount` implements `FormatForMode`.
In which case, the sentence becomes

```
Describes how a CliAccount [value] can be converted into a string ...
```

I do not think that the word "item" is commonly used to reference types.
I've seen it used for type values, but it is better to use the word "value" for that.  And in general, "item" is one of those words that are so ambiguous that it is often recommended they are not used at all when naming things.